### PR TITLE
RBAC negative tests

### PR DESCRIPTION
**Pull request by kacperpONS**
### What is the context of this PR?

Follow up to PR #128 adding more thorough unit tests including checking that the user groups **do not** have any additional, unexpected permissions.

### How to review

Run migrations and then: 
```
poetry run coverage run ./manage.py test --parallel --settings=cms.settings.test cms.core.tests.test_permissions --keepdb
```

### CODE REVIEW

File: cms/core/tests/test_permissions.py [hunk 1]
```diff
@@ -19,153 +131,208 @@ def create_user(cls, group_name):
         group = Group.objects.get(name=group_name)
         group.user_set.add(cls.user)
 
-    def all_permissions_check_helper(self, app: str, model: str):
-        """Helper method to check if a user has all permissions for a model."""
+        cls.user_permissions = list(cls.user.user_permissions.all() | Permission.objects.filter(group__user=cls.user))
+
+    def check_and_remove_from_user_permissions(self, app, model, permission_type):
```
Comment (MaciekBaron) [inline cms/core/tests/test_permissions.py:136]: Method missing type signatures.

I'm also wondering if it should use the existing convention and be renamed to have "helper" at the end?

Comment (kacperpONS) [reply cms/core/tests/test_permissions.py]: [2f92347](https://github.com/ONSdigital/dis-wagtail/pull/143/commits/2f92347cabf750b124022b0c223e1b3b74ffee08)

File: cms/core/tests/test_permissions.py [hunk 2]
```diff
@@ -1,16 +1,128 @@
...
 WAGTAIL_PERMISSION_TYPES = ["add", "change", "delete"]
 
 WAGTAIL_PAGE_PERMISSION_TYPES = ["add", "change", "bulk_delete", "lock", "publish", "unlock"]
 
 
-class PermissionsTestCase(TestCase):
+class TestPermissions(TestCase):
+    """Test edge cases for the mechanism used for adding permissions in the data migration."""
+
+    def test_raises_exception_when_app_doesnt_exist(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_permissions.py:27]: Nit:
```suggestion
    def test_exception_when_app_not_found(self):
```

File: cms/core/tests/test_permissions.py [hunk 3]
```diff
@@ -1,16 +1,128 @@
...
+        app = "non_existent_app"
+        model = "contactdetails"
+
+        permission_codename = "add_contactdetails"
+        group_name = settings.PUBLISHING_ADMINS_GROUP_NAME
+
+        with self.assertRaises(LookupError):
+            assign_permission_to_group(apps, group_name, permission_codename, app, model)
+
+    def test_raises_exception_when_model_doesnt_exist(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_permissions.py:38]: Nit:
```suggestion
    test_exception_when_model_not_found(self):
```

File: cms/core/tests/test_permissions.py [hunk 4]
```diff
@@ -1,16 +1,128 @@
...
+        """Check that an exception is raised when a model isn't registered in an app."""
+        app = "core"
+        model = "non_existent_model"
+        permission_codename = "add_contactdetails"
+        group_name = settings.PUBLISHING_ADMINS_GROUP_NAME
+
+        with self.assertRaises(LookupError):
+            assign_permission_to_group(apps, group_name, permission_codename, app, model)
+
+    def test_raises_exception_when_group_doesnt_exist(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_permissions.py:48]: Nit:
```suggestion
    test_exception_when_group_not_found(self):
```

File: cms/core/tests/test_permissions.py [hunk 5]
```diff
@@ -1,16 +1,128 @@
...
+        app = "core"
+        model = "contactdetails"
+        permission_codename = "add_contactdetails"
+        group_name = "non_existent_group"
+
+        with self.assertRaises(Group.DoesNotExist):
+            assign_permission_to_group(apps, group_name, permission_codename, app, model)
+
+    def test_user_can_add_model_when_has_permission(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_permissions.py:57]: Nit:
```suggestion
    test_add_model_with_permission(self):
```

File: cms/core/tests/test_permissions.py [hunk 6]
```diff
@@ -1,16 +1,128 @@
...
+
+        self.assertEqual(
+            response.status_code,
+            http.HTTPStatus.FOUND,
+        )
+
+        contact_details = ContactDetails.objects.get(name="Contact details")
+        self.assertIsNotNone(contact_details)
+
+    def test_user_cannot_add_model_when_doesnt_have_permission(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_permissions.py:88]: Nit:
```suggestion
    test_cannot_add_model_without_permission(self):
```

**Review by sanjeevz3009 — COMMENTED** [review]
Great work, thank you for addressing this/ doing the negative tests. `Least Privilege Principle` all over. Good stuff!

Couple minor suggestions. Other than that, everything looks good to me.

File: cms/core/tests/test_permissions.py (hunk 2 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_permissions.py]: Addressed!

File: cms/core/tests/test_permissions.py (hunk 4 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_permissions.py]: Addressed

File: cms/core/tests/test_permissions.py (hunk 6 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_permissions.py]: Addressed

File: cms/core/tests/test_permissions.py (hunk 5 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_permissions.py]: Addressed

File: cms/core/tests/test_permissions.py (hunk 3 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_permissions.py]: Addressed

**Review by sanjeevz3009 — APPROVED** [review]
LGTM!

**Review by AdamHawtin — APPROVED** [review]
Lots of valuable coverage 👍

**Review by MaciekBaron — APPROVED** [review]
LGTM!
